From 1c7b2e98f03cb813918d8b8b9967842ac4b657bb Mon Sep 17 00:00:00 2001
From: Baris Dinc <barisdinc@gmail.com>
Date: Sun, 11 Oct 2026 13:59:05 +0300
Subject: [PATCH 2/2] projects: pluto: Insert axi_datv_xpdr between RX and TX
 paths

Tap the RX I0/Q0 samples of axi_ad9361 (before the decimator) and
place axi_datv_xpdr between the TX interpolator output and the
axi_ad9361 DAC inputs. The core is mapped at 0x7C450000.

When the core is enabled and the DAC channel data select is set to
DMA (2), the TX output carries the processed RX samples. When disabled
the TX data path is unchanged.
---
 projects/pluto/Makefile      |  1 +
 projects/pluto/system_bd.tcl | 17 +++++++++++++++--
 2 files changed, 16 insertions(+), 2 deletions(-)

diff --git a/projects/pluto/Makefile b/projects/pluto/Makefile
index a8e40da6a..23f2a5fa6 100644
--- a/projects/pluto/Makefile
+++ b/projects/pluto/Makefile
@@ -18,6 +18,7 @@ M_DEPS += ../../library/axi_ad9361/axi_ad9361_delay.tcl
 LIB_DEPS += axi_ad9361
 LIB_DEPS += axi_dmac
 LIB_DEPS += axi_tdd
+LIB_DEPS += axi_datv_xpdr
 LIB_DEPS += util_pack/util_cpack2
 LIB_DEPS += util_pack/util_upack2
 
diff --git a/projects/pluto/system_bd.tcl b/projects/pluto/system_bd.tcl
index 4787085d0..4720b92cf 100644
--- a/projects/pluto/system_bd.tcl
+++ b/projects/pluto/system_bd.tcl
@@ -274,10 +274,22 @@ ad_connect axi_ad9361/l_clk tx_fir_interpolator/aclk
 
 ad_connect axi_ad9361/dac_enable_i0 tx_fir_interpolator/dac_enable_0
 ad_connect axi_ad9361/dac_valid_i0 tx_fir_interpolator/dac_valid_0
-ad_connect axi_ad9361/dac_data_i0 tx_fir_interpolator/data_out_0
 ad_connect axi_ad9361/dac_enable_q0 tx_fir_interpolator/dac_enable_1
 ad_connect axi_ad9361/dac_valid_q0 tx_fir_interpolator/dac_valid_1
-ad_connect axi_ad9361/dac_data_q0 tx_fir_interpolator/data_out_1
+
+# datv repeater: RX I0/Q0 -> gain/squelch -> TX I0/Q0 (bypassed when disabled)
+
+ad_ip_instance axi_datv_xpdr datv_xpdr
+
+ad_connect axi_ad9361/l_clk datv_xpdr/clk
+ad_connect axi_ad9361/rst datv_xpdr/rst
+ad_connect axi_ad9361/adc_valid_i0 datv_xpdr/rx_valid
+ad_connect axi_ad9361/adc_data_i0 datv_xpdr/rx_data_i
+ad_connect axi_ad9361/adc_data_q0 datv_xpdr/rx_data_q
+ad_connect tx_fir_interpolator/data_out_0 datv_xpdr/tx_in_data_i
+ad_connect tx_fir_interpolator/data_out_1 datv_xpdr/tx_in_data_q
+ad_connect datv_xpdr/tx_out_data_i axi_ad9361/dac_data_i0
+ad_connect datv_xpdr/tx_out_data_q axi_ad9361/dac_data_q0
 
 ad_connect  axi_ad9361/l_clk tx_upack/clk
 
@@ -353,6 +365,7 @@ ad_cpu_interconnect 0x7C400000 axi_ad9361_adc_dma
 ad_cpu_interconnect 0x7C420000 axi_ad9361_dac_dma
 ad_cpu_interconnect 0x7C430000 axi_spi
 ad_cpu_interconnect 0x7C440000 axi_tdd_0
+ad_cpu_interconnect 0x7C450000 datv_xpdr
 
 ad_ip_parameter sys_ps7 CONFIG.PCW_USE_S_AXI_HP1 {1}
 ad_connect sys_cpu_clk sys_ps7/S_AXI_HP1_ACLK
-- 
2.43.0

